app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsPagingSource.kt

```diff
@@ -0,0 +1,184 @@
+ *
+ * This file is a part of Tusky.
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>.
+ */
+
+package com.keylesspalace.tusky.components.notifications
+
+import android.util.Log
+import androidx.paging.PagingSource
+import androidx.paging.PagingState
+import com.keylesspalace.tusky.entity.Notification
+import com.keylesspalace.tusky.network.MastodonApi
+import com.keylesspalace.tusky.util.HttpHeaderLink
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import okhttp3.Headers
+import retrofit2.Response
+import javax.inject.Inject
+
+/** Models next/prev links from the "Links" header in an API response */
+data class Links(val next: String?, val prev: String?)
+
+/** [PagingSource] for Mastodon Notifications, identified by the Notification ID */
+class NotificationsPagingSource @Inject constructor(
+    private val mastodonApi: MastodonApi,
+    private val notificationFilter: Set<Notification.Type>
+) : PagingSource<String, Notification>() {
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, Notification> {
+        Log.d(TAG, "load() with ${params.javaClass.simpleName} for key: ${params.key}")
+
+        try {
+            val response = when (params) {
+                is LoadParams.Refresh -> {
+                    getInitialPage(params)
+                }
+                is LoadParams.Append -> mastodonApi.notifications(
+                    maxId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+                is LoadParams.Prepend -> mastodonApi.notifications(
+                    minId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+            }
+
+            if (!response.isSuccessful) {
+                return LoadResult.Error(Throwable(response.errorBody().toString()))
+            }
+
+            val links = getPageLinks(response.headers()["link"])
+            return LoadResult.Page(
+                data = response.body()!!,
+                nextKey = links.next,
+                prevKey = links.prev
+            )
+        } catch (e: Exception) {
+            return LoadResult.Error(e)
+        }
+    }
+
+    /**
+     * Fetch the initial page of notifications, using params.key as the ID of the initial
```

Comment:
That key could be long in the past? (Ie thousands of notifications since then)

Reply:
I've not been able to think of a plausible scenario where that could happen. I'll resolve this, but please re-open if you can describe one.